File: atlas/examples/kotlin/domain/src/main/kotlin/dev/atlas/example/domain/catalog/CatalogItem.kt
package dev.atlas.example.domain.catalog

/**
 * Defines the common identity and descriptive behavior of every catalog item.
 *
 * @property id Stable non-empty catalog identifier.
 * @property title Normalized human-readable title.
 * @property kind Concrete item specialization.
 */
abstract class CatalogItem(
    val id: CatalogItemId,
    val title: String,
    val kind: CatalogItemKind
) {
    /**
     * Describes the specialized catalog content.
     *
     * @return Concise user-facing item description.
     */
    abstract fun describe(): String
}

File: atlas/examples/typescript/packages/domain/src/catalog/CatalogItem.ts
import type { CatalogItemId } from './CatalogItemId.js';
import type { CatalogItemKind } from './CatalogItemKind.js';

/**
 * Defines the common identity and descriptive behavior of every catalog item.
 */
export abstract class CatalogItem {
  /**
   * Creates an item from values already normalized by the domain factory.
   *
   * @param id - Stable non-empty catalog identifier.
   * @param title - Normalized human-readable title.
   * @param kind - Concrete item specialization.
   */
  protected constructor(
    public readonly id: CatalogItemId,
    public readonly title: string,
    public readonly kind: CatalogItemKind,
  ) {}

  /**
   * Describes the specialized catalog content.
   *
   * @returns Concise user-facing item description.
   */
  public abstract describe(): string;
}
